faunadb-java-dsl/src/test/java/com/faunadb/client/DecoderSpec.java

```diff
@@ -0,0 +1,490 @@
+        obj.setMapStrToStr(newHashMap(of("key", "value")));
+
+        ObjectV objectV = new ObjectV(of(
+                "strField", new StringV("value"),
+                "longField", new LongV(10),
+                "listStrField", new ArrayV(asList(new StringV("value1"), new StringV("value2"))),
+                "setLongField", new ArrayV(asList(new LongV(1), new LongV(2), new LongV(3))),
+                "mapStrToStr", new ObjectV(of("key", new StringV("value")))
+        ));
+        assertEquals(obj, decode(objectV, SimpleObject.class));
+    }
+
+    static class ObjectWithConstructor {
+        private final String strField;
+
+        @FaunaConstructor
+        public ObjectWithConstructor(@FaunaField("strField") String strField) {
+            this.strField = strField;
+        }
+
+        public String getStrField() {
+            return strField;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            ObjectWithConstructor other = (ObjectWithConstructor) obj;
+            return strField.equals(other.strField);
+        }
+
+        @Override
+        public String toString() {
+            return format("%s(%s)", getClass().getSimpleName(), strField);
+        }
+    }
+
+    @Test
+    public void shouldDecodeObjectsWithConstructor() {
+        ObjectWithConstructor obj = new ObjectWithConstructor("value");
+
+        ObjectV objectV = new ObjectV(of("strField", new StringV("value")));
+
+        assertEquals(obj, decode(objectV, ObjectWithConstructor.class));
+    }
+
+    static class ObjectWithStaticCreator {
+        private final String strField;
+        private final long longField;
+
+        ObjectWithStaticCreator(String strField, long longField) {
+            this.strField = strField;
+            this.longField = longField;
+        }
+
+        @FaunaConstructor
+        public static Object creator(@FaunaField("strField") String strField) {
+            return new ObjectWithStaticCreator(strField, 10);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            ObjectWithStaticCreator other = (ObjectWithStaticCreator) obj;
+            return strField.equals(other.strField) && longField == other.longField;
+        }
+
+        @Override
+        public String toString() {
+            return format("%s(%s, %s)", getClass().getSimpleName(), strField, longField);
+        }
+    }
+
+    @Test
+    public void shouldDecodeObjectsWithStaticCreator() {
+        ObjectWithStaticCreator obj = new ObjectWithStaticCreator("value", 10);
+
+        ObjectV objectV = new ObjectV(of("strField", new StringV("value")));
+
+        assertEquals(obj, decode(objectV, ObjectWithStaticCreator.class));
+    }
+
+    static class ObjectWithCreatorAndPropertiesMixed {
+        private String strField;
+        private long longField;
+
+        @FaunaConstructor
+        public static Object creator(@FaunaField("strField") String strField) {
```

Comment:
What if the `longField` was part of the constructor too? Does it call the constructor and the setter?